fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectory.java

```diff
@@ -596,25 +597,36 @@ public void deleteFile(@Nonnull String name) throws IOException {
             return;
         }
         try {
-            boolean deleted = Objects.requireNonNull(context.asyncToSync(LuceneEvents.Waits.WAIT_LUCENE_DELETE_FILE, getFileReferenceCacheAsync().thenApply(cache -> {
-                FDBLuceneFileReference value = cache.get(name);
-                if (value == null) {
-                    return false;
-                }
-                context.ensureActive().clear(metaSubspace.pack(name));
-                context.ensureActive().clear(dataSubspace.subspace(Tuple.from(value.getId())).range());
-                cache.remove(name);
-                return true;
-            })));
+            boolean deleted = Objects.requireNonNull(context.asyncToSync(LuceneEvents.Waits.WAIT_LUCENE_DELETE_FILE, getFileReferenceCacheAsync()
+                    .thenApply(cache -> deleteFileInternal(cache, name))));
 
             if (!deleted) {
                 throw new NoSuchFileException(name);
             }
+
+            if (isCompoundFile(name)) {
+                Map<String, FDBLuceneFileReference> cache = fileReferenceCache.get();
+                String primaryKeyName = name.substring(0, name.length() - DATA_EXTENSION.length()) + "pky";
+                deleteFileInternal(cache, primaryKeyName);
+                // TODO: If the segment is being deleted because it no longer has any live docs, it won't be merged
+                //  so we might need to delete the primary key entries another way for case where tryDeleteDocument
+                //  wasn't used.
```

Comment:
The trick that I used in https://github.com/FoundationDB/fdb-record-layer/pull/2300/files/f55d9a8ba7c7f449b30463ea5857ea08c47bd096..63fd482f1af4b11c401ed05c92b9c8c80605603b (which has yet to be reviewed, so maybe someone will point out it's a _bad_ idea)
Is to have files that are actually tracked in the filesystem, but have custom delete logic.
This would basically mean that `LucenePrimaryKeySegmentIndex.WrappedFieldsWriter` would open an additional `*.pky` for the segment when writing, etc. Lucene would then now that the file is part of the segment, and when it goes to delete it, it would send the delete to the `FDBDirectory`. This could then call back into `LucenePrimaryKeySegmentIndex` to delete any entries for that segment. That being said, I think you would have to ensure that the `*.pky` is delete _after_ the stored fields data, so that you can efficiently find the primary keys that you have to delete.
We could, instead, have the `LucenePrimaryKeySegmentIndex` baked into our `StoredFieldsFormat` and then when you delete the data for `StoredFields` we will delete the associated `*.pky` data.

Reply:
I was talking with @ohadzeliger offline about this. When the documents are deleted we immediately delete the entries from the index, so the issue here would only be that the `*.pky` file reference is left around, and that needs to get cleaned up, right?

Reply:
Yes, currently. Though we might want to use the lifetime of the `.pky` file reference to trigger some additional cleanup if we determine that there are missing cases with the inline one.